backends/tc/introspection.h

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright (C) 2023 Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions
+and limitations under the License.
+*/
+
+#ifndef BACKENDS_TC_INTROSPECTION_H_
+#define BACKENDS_TC_INTROSPECTION_H_
+
+#include "frontends/p4/parseAnnotations.h"
+#include "frontends/p4/parserCallGraph.h"
+#include "ir/ir.h"
+#include "lib/json.h"
+#include "lib/nullstream.h"
+#include "options.h"
+
+namespace TC {
+
+struct introspectionInfo {
+    cstring schemaVersion;
+    cstring pipelineName;
+    unsigned int pipelineId;
+    introspectionInfo() {
+        schemaVersion = nullptr;
+        pipelineName = nullptr;
+        pipelineId = 0;
+    }
+    void initIntrospectionInfo(IR::TCPipeline *tcPipeline) {
+        schemaVersion = "1.0.0";
+        pipelineName = tcPipeline->pipelineName;
+        pipelineId = tcPipeline->pipelineId;
+    }
+};
+
+struct keyFieldAttributes {
+    unsigned int id;
+    cstring name;
+    cstring type;
+    cstring matchType;
+    unsigned int bitwidth;
+    keyFieldAttributes() {
+        id = 0;
+        name = nullptr;
+        type = nullptr;
+        matchType = nullptr;
+        bitwidth = 0;
+    }
+};
+
+struct annotation {
+    cstring name;
+    annotation() { name = nullptr; }
+    explicit annotation(cstring n) { name = n; }
+};
+
+struct actionParam {
+    unsigned int id;
+    cstring name;
+    unsigned int dataType;
+    unsigned int bitwidth;
+    actionParam() {
+        id = 0;
+        name = nullptr;
+        bitwidth = 0;
+    }
+};
+
+enum actionScope { TableOnly, DefaultOnly, TableAndDefault };
+
+struct actionAttributes {
+    unsigned int id;
+    cstring name;
+    actionScope scope;
+    bool defaultHit;
+    bool defaultMiss;
+    safe_vector<struct annotation *> annotations;
+    safe_vector<struct actionParam *> actionParams;
+    actionAttributes() {
+        id = 0;
+        name = nullptr;
+        scope = TableAndDefault;
+        defaultHit = false;
+        defaultMiss = false;
+    }
+};
+
+struct tableAttributes {
+    cstring name;
+    unsigned int id;
+    unsigned int tentries;
+    unsigned int numMask;
+    unsigned int keysize;
+    unsigned int keyid;
+    safe_vector<struct keyFieldAttributes *> keyFields;
+    safe_vector<struct actionAttributes *> actions;
+    tableAttributes() {
+        name = nullptr;
+        id = 0;
+        tentries = 0;
+        numMask = 0;
+        keysize = 0;
+        keyid = 0;
+    }
+};
+
+// This pass generates introspection JSON into user specified file
```

Comment:
this should be a triple slash.
Who does the introspection?

Reply:
introspection is done by control plane